src/libopensc/card-esteid2018.c

```diff
@@ -0,0 +1,327 @@
+		if (memcmp(path, "\x3F\x00", 2) == 0) {
+			LOG_TEST_RET(card->ctx, esteid_select(card, 0x00, 0x3F, 0x00), "MF select failed");
+		} else if (pathlen >= 2 && path[0] == 0xAD) {
+			LOG_TEST_RET(card->ctx, esteid_select(card, 0x01, path[0], path[1]), "DF select failed");
+		} else if (pathlen == 2) {
+			LOG_TEST_RET(card->ctx, esteid_select(card, 0x02, path[0], path[1]), "EF select failed");
+
+			if (file_out != NULL) // Just make a dummy file
+			{
+				file = sc_file_new();
+				if (file == NULL)
+					LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+				file->path = *in_path;
+
+				*file_out = file;
+			}
+		}
+		path += 2;
+		pathlen -= 2;
+	}
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+// temporary hack, overload 6B00 SW processing
+static int esteid_read_binary(struct sc_card *card, unsigned int idx, u8 *buf, size_t count, unsigned long flags) {
+	int r;
+	int (*saved)(struct sc_card *, unsigned int, unsigned int) = card->ops->check_sw;
+	LOG_FUNC_CALLED(card->ctx);
+	card->ops->check_sw = esteid_check_sw;
+	r = iso_ops->read_binary(card, idx, buf, count, flags);
+	card->ops->check_sw = saved;
+	LOG_FUNC_RETURN(card->ctx, r);
+}
+
+static int esteid_set_security_env(sc_card_t *card, const sc_security_env_t *env, int se_num) {
+	struct esteid_priv_data *priv;
+	struct sc_apdu apdu;
+
+	// XXX: could be const
+	unsigned char cse_crt_aut[] = {0x80, 0x04, 0xFF, 0x20, 0x08, 0x00, 0x84, 0x01, 0x81};
+	unsigned char cse_crt_sig[] = {0x80, 0x04, 0xFF, 0x15, 0x08, 0x00, 0x84, 0x01, 0x9F};
+	unsigned char cse_crt_dec[] = {0x80, 0x04, 0xFF, 0x30, 0x04, 0x00, 0x84, 0x01, 0x81};
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	if (card == NULL || env == NULL || env->key_ref_len != 1)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+	sc_log(card->ctx, "algo: %d operation: %d keyref: %d", env->algorithm, env->operation, env->key_ref[0]);
+
+	if (env->algorithm == SC_ALGORITHM_EC && env->operation == SC_SEC_OPERATION_SIGN && env->key_ref[0] == 1) {
+		sc_format_apdu_ex(card, &apdu, 0x22, 0x41, 0xA4, cse_crt_aut, sizeof(cse_crt_aut), NULL, 0);
+	} else if (env->algorithm == SC_ALGORITHM_EC && env->operation == SC_SEC_OPERATION_SIGN && env->key_ref[0] == 2) {
+		sc_format_apdu_ex(card, &apdu, 0x22, 0x41, 0xB6, cse_crt_sig, sizeof(cse_crt_sig), NULL, 0);
+	} else if (env->algorithm == SC_ALGORITHM_EC && env->operation == SC_SEC_OPERATION_DERIVE && env->key_ref[0] == 1) {
+		sc_format_apdu_ex(card, &apdu, 0x22, 0x41, 0xB8, cse_crt_dec, sizeof(cse_crt_dec), NULL, 0);
+	} else {
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_SUPPORTED);
+	}
+	SC_TRANSMIT_TEST_RET(card, apdu, "SET SECURITY ENV failed");
+
+	priv = DRVDATA(card);
+	priv->sec_env = *env;
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int esteid_compute_signature(sc_card_t *card, const u8 *data, size_t datalen, u8 *out, size_t outlen) {
+	struct esteid_priv_data *priv = DRVDATA(card);
+	struct sc_security_env *env = NULL;
+	struct sc_apdu apdu;
+	u8 sbuf[SIGNATURE_PAYLOAD_SIZE] = {0};
```

Comment:
Either use `= {0};` or `memset(sbuf, 0x00, SIGNATURE_PAYLOAD_SIZE - datalen);`, but don't use both. The former should work starting with C99.